llvm/test/Analysis/LoopAccessAnalysis/non-constant-strides-backward.ll

```diff
@@ -8,15 +8,21 @@ declare void @llvm.assume(i1)
 define void @different_non_constant_strides_known_backward(ptr %A) {
 ; CHECK-LABEL: 'different_non_constant_strides_known_backward'
 ; CHECK-NEXT:    loop:
-; CHECK-NEXT:      Report: unsafe dependent memory operations in loop. Use #pragma clang loop distribute(enable) to allow loop distribution to attempt to isolate the offending operations into a separate loop
-; CHECK-NEXT:  Unknown data dependence.
+; CHECK-NEXT:      Memory dependences are safe with run-time checks
```

Comment:
I'm not sure the resulting analysis is correct for this loop now. Forcing the target instruction cost in LoopVec results in vectorization with a VF of 4 (at least for NEON) with the following RT checks:

```
vector.memcheck:                                  ; preds = %entry
  %scevgep = getelementptr i8, ptr %A, i64 2044
  %scevgep1 = getelementptr i8, ptr %A, i64 1024
  %bound0 = icmp ult ptr %A, %scevgep1
  %bound1 = icmp ult ptr %A, %scevgep
  %found.conflict = and i1 %bound0, %bound1
  br i1 %found.conflict, label %scalar.ph, label %vector.ph
```

If I'm right, we're safe in this instance because the check should always fail (unless the scevgeps wrap), and we never enter the vectorized code because `%A` should be lower than both `%A+1024` and `%A+2044`. So we've just bloated our binary a bit without a payoff, but should still get the right result.

If at least one scevgep wraps and we enter the vectorized code, then I think we'll get a miscompare with a VF above 2, since you'll be reading from `%A+2` when that location would be written to one lane earlier in the scalar loop:
Scalar iteration 1:
read `%A+0`, write `%A+0`
Scalar iteration 2:
read `%A+1`, write `%A+2`
Scalar iteration 3, hazard with static VF>2:
read `%A+2`, write `%A+4`

Please let me know if I've gotten mixed up here.

Reply:
I think you're right: I didn't check the actual runtime checks generated. LAA prints the runtime checks correctly:

```
; CHECK-NEXT:      Run-time memory checks:
; CHECK-NEXT:      Check 0:
; CHECK-NEXT:        Comparing group ([[GRP1:0x[0-9a-f]+]]):
; CHECK-NEXT:          %gep.mul.2 = getelementptr inbounds i32, ptr %A, i64 %iv.mul.2
; CHECK-NEXT:        Against group ([[GRP2:0x[0-9a-f]+]]):
; CHECK-NEXT:          %gep = getelementptr inbounds i32, ptr %A, i64 %iv
```

For some reason, the RT-check generation routine forgets to add the `inbounds` to the GEPs, and I wonder if this is an underlying bug in LAA.

Another issue is that, we seem to be overly aggressive about generating runtime checks that we're not able to prove are always false (perhaps with `inbounds`, it can be proven?).

Will investigate in more detail, and take a fresh look at the patch, thanks!

Reply:
Turns out this is due to some ongoing work in SCEVExpander and GEPNoWrapFlags. The scevgeps should now be marked nuw, as per https://discourse.llvm.org/t/rfc-add-nusw-and-nuw-flags-for-getelementptr/78672.